File: py/herdr_run/readiness.py
"""Decide whether a pane is safe to type a command into.

This is the dangerous part of the tool, so it is built out of two INDEPENDENT signals that must
agree, and it fails closed.

**Primary — the foreground process group (an observable, not a proxy).** ``herdr pane
process-info`` reports the pane's ``shell_pid`` and its ``foreground_process_group_id``. When the
shell is at its prompt, the foreground process group IS the shell: the pgid equals the shell pid and
the only foreground process is the shell itself. When anything is running, the kernel has moved the
foreground group to the job. This dereferences the actual running process rather than inferring
liveness from what the screen looks like, so it does not care about prompt themes, colour, spinners,
or how a TUI happens to redraw.

**Secondary — the prompt tail (a veto, never an authority).** The process signal cannot see a
command a human TYPED without pressing Enter: the shell is genuinely idle, yet typing into that
pane would concatenate our command onto their half-written line. So we also look at the last
rendered line. We infer the prompt's trailing literal from the shell rc file (for a ``PS1`` ending
``...\\n\\$ `` the tail is ``"$ "``), then:

* last line ends with the tail            -> ``clean``
* tail occurs earlier in the line         -> ``dirty``  (something is typed after the prompt)
* tail not found, or could not be inferred -> ``abstain``

``abstain`` is honest rather than optimistic: this signal only ever VETOES on positive evidence of
dirt, and never upgrades a pane to ready on its own. If it could not understand the prompt it says
so, and the record shows ``prompt=abstain`` instead of implying a check that did not happen.

A pane is READY only when the process signal says idle and the prompt signal does not veto — and,
because a command being launched has a brief window where the pgid has not moved yet, the process
signal must hold for two consecutive polls.
"""

from __future__ import annotations

import os
import re
from dataclasses import dataclass

from herdr_run.client import HerdrClient, ProcessInfo
from herdr_run.config import Config

__all__ = ["ProcessSignal", "PromptSignal", "Readiness", "assess", "infer_prompt_tail", "parse_ps1"]

#: ``\[ ... \]`` marks zero-width (non-printing) sequences in a bash PS1 — colour codes and the
#: like. They render to nothing, so they must be removed before asking what the prompt LOOKS like.
_ZERO_WIDTH = re.compile(r"\\\[.*?\\\]")

#: A backslash escape (``\u``, ``\h``, ``\w``, ``\$``, ...) or a shell expansion (``$(...)``,
#: ``${...}``, ``$VAR``). Everything after the LAST of these is fixed literal text, which is exactly
#: the trailing run we can match against a rendered line.
_DYNAMIC = re.compile(r"\\.|\$\([^)]*\)|\$\{[^}]*\}|\$[A-Za-z_][A-Za-z0-9_]*")

#: ``PS1='...'`` / ``PS1="..."`` / ``PROMPT=...`` assignment at the start of a line (optionally
#: `export`ed). Deliberately anchored so a commented-out example (`# PS1=...`) is not picked up.
_PS1_ASSIGN = re.compile(
    r"^[ \t]*(?:export[ \t]+)?(?:PS1|PROMPT)=(\"(?:[^\"\\]|\\.)*\"|'[^']*'|[^\s#]+)[ \t]*(?:#.*)?$",
    re.MULTILINE,
)


@dataclass(frozen=True)
class ProcessSignal:
    """What the pane's foreground process group says about whether the shell is at a prompt."""

    idle: bool
    reason: str
    shell_pid: int
    foreground_pgid: int
    foreground: tuple[str, ...]


@dataclass(frozen=True)
class PromptSignal:
    """What the last rendered line says. A veto only: it never declares a pane ready."""

    #: One of ``clean``, ``dirty``, ``abstain``.
    verdict: str
    reason: str
    tail: str | None
    last_line: str | None


@dataclass(frozen=True)
class Readiness:
    """The combined verdict of both signals, and the reasoning behind it."""

    ready: bool
    reason: str
    process: ProcessSignal
    prompt: PromptSignal

    def describe(self) -> str:
        """One line naming both signals and their reasons, for logs and error messages."""
        return (
            f"process={'idle' if self.process.idle else 'busy'} ({self.process.reason}); "
            f"prompt={self.prompt.verdict} ({self.prompt.reason})"
        )


def parse_ps1(rc_text: str) -> str | None:
    """Return the LAST uncommented ``PS1``/``PROMPT`` assignment in a shell rc file, unquoted.

    Last wins because that is what the shell itself ends up with when an rc file reassigns the
    prompt inside a conditional branch further down.
    """
    # `findall` is typed as returning Any; take the capture group off the match objects instead so
    # the value stays a concrete `str` under `disallow_any_explicit`.
    raw: str | None = None
    for match in _PS1_ASSIGN.finditer(rc_text):
        raw = match.group(1)
    if raw is None:
        return None
    if len(raw) >= 2 and raw[0] == raw[-1] and raw[0] in "\"'":
        return raw[1:-1]
    return raw


def prompt_tail_of(ps1: str) -> str | None:
    """Extract the fixed literal text a rendered prompt ends with, or ``None`` if there is none.

    ``"\\[\\033[0;34m\\][\\u@\\h \\w] \\n\\$ \\[\\033[0m\\]"`` -> ``"$ "``.
    """
    visible = _ZERO_WIDTH.sub("", ps1)
    last_dynamic_end = 0
    for match in _DYNAMIC.finditer(visible):
        # `\$` renders as a literal `$` (or `#` for root) and is the single most common prompt
        # terminator, so treat it as part of the tail rather than as an opaque expansion.
        if match.group(0) == r"\$":
            last_dynamic_end = match.start()
            continue
        last_dynamic_end = match.end()
    tail = visible[last_dynamic_end:]
    tail = tail.replace(r"\$", "$")
    return tail if tail.strip() else None


def infer_prompt_tail(config: Config, *, home: str | None = None) -> str | None:
    """Resolve the prompt tail from explicit config, else by reading the user's shell rc files."""
    if config.prompt_tail is not None:
        return config.prompt_tail
    base = home if home is not None else os.path.expanduser("~")
    for name in (".bashrc", ".zshrc", ".bash_profile", ".profile"):
        path = os.path.join(base, name)
        try:
            with open(path, encoding="utf-8", errors="replace") as handle:
                text = handle.read()
        except OSError:
            continue
        ps1 = parse_ps1(text)
        if ps1 is None:
            continue
        tail = prompt_tail_of(ps1)
        if tail is not None:
            return tail
    return None


def assess_process(info: ProcessInfo, config: Config) -> ProcessSignal:
    """Judge idleness from the foreground process group. Idle only when the shell owns it alone."""
    names = tuple(name for _pid, name, _cmdline in info.foreground)
    common = ProcessSignal(
        idle=False,
        reason="",
        shell_pid=info.shell_pid,
        foreground_pgid=info.foreground_pgid,
        foreground=names,
    )
    if info.foreground_pgid != info.shell_pid:
        running = ", ".join(names) or "unknown"
        return _with_reason(
            common,
            False,
            f"foreground pgid {info.foreground_pgid} != shell pid {info.shell_pid}; running: {running}",
        )
    if len(info.foreground) != 1:
        return _with_reason(common, False, f"{len(info.foreground)} foreground processes, expected only the shell")
    pid, name, _cmdline = info.foreground[0]
    if pid != info.shell_pid:
        return _with_reason(common, False, f"foreground process {pid} ({name}) is not the shell {info.shell_pid}")
    if name not in config.shells:
        # A pane whose shell slot is occupied by something else (an editor, an agent TUI, a REPL) is
        # not a shell prompt, whatever its process group says.
        return _with_reason(common, False, f"foreground process is {name!r}, not a known shell")
    return _with_reason(common, True, f"shell {name} ({info.shell_pid}) owns the foreground group")


def _with_reason(signal: ProcessSignal, idle: bool, reason: str) -> ProcessSignal:
    return ProcessSignal(
        idle=idle,
        reason=reason,
        shell_pid=signal.shell_pid,
        foreground_pgid=signal.foreground_pgid,
        foreground=signal.foreground,
    )


def assess_prompt(text: str, tail: str | None) -> PromptSignal:
    """Judge the last rendered line: ``clean``, ``dirty``, or ``abstain`` when it is unreadable."""
    if tail is None:
        return PromptSignal("abstain", "no prompt tail configured or inferable", None, None)
    lines = [line for line in text.splitlines() if line.strip()]
    if not lines:
        return PromptSignal("abstain", "pane has no rendered output yet", tail, None)
    last = lines[-1]
    # `herdr pane read` strips trailing whitespace from each line, so compare against the stripped
    # tail; an idle prompt of "$ " renders as the line "$".
    needle = tail.rstrip()
    if not needle:
        return PromptSignal("abstain", "prompt tail is whitespace only", tail, last)
    if last.rstrip().endswith(needle):
        return PromptSignal("clean", f"last line ends with prompt tail {needle!r}", tail, last)
    if needle in last:
        return PromptSignal("dirty", f"text typed after prompt tail {needle!r}: {last.rstrip()[-80:]!r}", tail, last)
    return PromptSignal("abstain", f"prompt tail {needle!r} not found in last line", tail, last)


def assess(
    client: HerdrClient,
    pane_id: str,
    config: Config,
    *,
    prompt_tail: str | None,
    read_lines: int = 4,
) -> Readiness:
    """Take one readiness reading of ``pane_id``. Callers poll this; it does not sleep."""
    process = assess_process(client.process_info(pane_id), config)
    if config.readiness == "process":
        prompt = PromptSignal("abstain", "prompt check disabled (readiness: process)", prompt_tail, None)
    else:
        prompt = assess_prompt(client.read(pane_id, lines=read_lines), prompt_tail)

    if not process.idle:
        return Readiness(False, "pane is busy", process, prompt)
    if prompt.verdict == "dirty":
        return Readiness(False, "pane has an unsubmitted command line", process, prompt)
    return Readiness(True, "pane is idle at a shell prompt", process, prompt)

File: rs/herdr-run/src/readiness.rs
//! Fail-closed readiness checks for a Herdr pane.
//!
//! The foreground process group is authoritative: an idle shell must own the terminal's
//! foreground group alone.  Rendered prompt text is an independent veto for a human's
//! unsubmitted command line.  Failure to infer a prompt is reported as abstention rather than
//! pretending that the text signal succeeded.

use std::fs;
use std::path::Path;
use std::sync::OnceLock;

use regex::Regex;

use crate::client::{current_account_home, HerdrApi, ProcessInfo};
use crate::config::Config;
use crate::error::Result;

/// Foreground-process evidence about whether the pane's shell is idle.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ProcessSignal {
    /// True only when the known shell alone owns the foreground process group.
    pub idle: bool,
    /// Human-readable explanation of the verdict.
    pub reason: String,
    /// PID Herdr reports for the interactive shell.
    pub shell_pid: i64,
    /// Foreground process-group ID Herdr reports for the pane.
    pub foreground_pgid: i64,
    /// Names of the processes in the foreground process group.
    pub foreground: Vec<String>,
}

/// Prompt-text evidence, used only to veto an otherwise idle pane.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct PromptSignal {
    /// One of `clean`, `dirty`, or `abstain`.
    pub verdict: String,
    /// Human-readable explanation of the verdict.
    pub reason: String,
    /// Prompt suffix searched for, if one was available.
    pub tail: Option<String>,
    /// Last nonblank rendered line, if one was available.
    pub last_line: Option<String>,
}

/// Combined process and prompt-text readiness evidence.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct Readiness {
    /// True only when the process signal is idle and prompt text does not veto it.
    pub ready: bool,
    /// Short combined-verdict explanation.
    pub reason: String,
    /// Foreground-process observation.
    pub process: ProcessSignal,
    /// Prompt-text observation.
    pub prompt: PromptSignal,
}

impl Readiness {
    /// Render one audit-friendly line naming both signals and their explanations.
    #[must_use]
    pub fn describe(&self) -> String {
        format!(
            "process={} ({}); prompt={} ({})",
            if self.process.idle { "idle" } else { "busy" },
            self.process.reason,
            self.prompt.verdict,
            self.prompt.reason
        )
    }
}

/// Return the last uncommented `PS1` or `PROMPT` assignment in shell RC text.
#[must_use]
pub fn parse_ps1(rc_text: &str) -> Option<String> {
    let mut raw = None;
    for captures in ps1_assignment().captures_iter(rc_text) {
        raw = captures.get(1).map(|value| value.as_str().to_owned());
    }
    let raw = raw?;
    let bytes = raw.as_bytes();
    if bytes.len() >= 2
        && bytes.first() == bytes.last()
        && matches!(bytes.first(), Some(b'\'' | b'"'))
    {
        Some(raw[1..raw.len() - 1].to_owned())
    } else {
        Some(raw)
    }
}

/// Extract the fixed literal suffix that a rendered prompt ends with.
#[must_use]
pub fn prompt_tail_of(ps1: &str) -> Option<String> {
    let visible = zero_width().replace_all(ps1, "");
    let mut last_dynamic_end = 0;
    for matched in dynamic().find_iter(&visible) {
        // `\$` is the common literal prompt terminator, so retain it in the fixed suffix.
        last_dynamic_end = if matched.as_str() == r"\$" {
            matched.start()
        } else {
            matched.end()
        };
    }
    let tail = visible[last_dynamic_end..].replace(r"\$", "$");
    (!tail.trim().is_empty()).then_some(tail)
}

/// Resolve the configured prompt tail, or infer it from account shell startup files.
///
/// Supplying `home` is intended for deterministic tests.  Production callers pass `None`, which
/// reads the current user's account-database home and never consults caller-controlled `HOME`.
#[must_use]
pub fn infer_prompt_tail(config: &Config, home: Option<&Path>) -> Option<String> {
    if let Some(tail) = &config.prompt_tail {
        return Some(tail.clone());
    }
    let account_home;
    let base = match home {
        Some(home) => home,
        None => {
            account_home = current_account_home().ok()?;
            &account_home
        }
    };
    for name in [".bashrc", ".zshrc", ".bash_profile", ".profile"] {
        let Ok(bytes) = fs::read(base.join(name)) else {
            continue;
        };
        let text = String::from_utf8_lossy(&bytes);
        let Some(ps1) = parse_ps1(&text) else {
            continue;
        };
        if let Some(tail) = prompt_tail_of(&ps1) {
            return Some(tail);
        }
    }
    None
}

/// Judge pane idleness from the foreground process group.
#[must_use]
pub fn assess_process(info: &ProcessInfo, config: &Config) -> ProcessSignal {
    let names = info
        .foreground
        .iter()
        .map(|(_, name, _)| name.clone())
        .collect::<Vec<_>>();
    let signal = |idle: bool, reason: String| ProcessSignal {
        idle,
        reason,
        shell_pid: info.shell_pid,
        foreground_pgid: info.foreground_pgid,
        foreground: names.clone(),
    };
    if info.foreground_pgid != info.shell_pid {
        let running = if names.is_empty() {
            "unknown".to_owned()
        } else {
            names.join(", ")
        };
        return signal(
            false,
            format!(
                "foreground pgid {} != shell pid {}; running: {running}",
                info.foreground_pgid, info.shell_pid
            ),
        );
    }
    if info.foreground.len() != 1 {
        return signal(
            false,
            format!(
                "{} foreground processes, expected only the shell",
                info.foreground.len()
            ),
        );
    }
    let (pid, name, _) = &info.foreground[0];
    if *pid != info.shell_pid {
        return signal(
            false,
            format!(
                "foreground process {pid} ({name}) is not the shell {}",
                info.shell_pid
            ),
        );
    }
    if !config.shells.iter().any(|shell| shell == name) {
        return signal(
            false,
            format!("foreground process is {name:?}, not a known shell"),
        );
    }
    signal(
        true,
        format!(
            "shell {name} ({}) owns the foreground group",
            info.shell_pid
        ),
    )
}

/// Judge the last rendered line as `clean`, `dirty`, or `abstain`.
#[must_use]
pub fn assess_prompt(text: &str, tail: Option<&str>) -> PromptSignal {
    let Some(tail) = tail else {
        return prompt_signal(
            "abstain",
            "no prompt tail configured or inferable",
            None,
            None,
        );
    };
    let last = text
        .lines()
        .rfind(|line| !line.trim().is_empty())
        .map(str::to_owned);
    let Some(last) = last else {
        return prompt_signal(
            "abstain",
            "pane has no rendered output yet",
            Some(tail),
            None,
        );
    };
    // Herdr strips trailing spaces from rendered lines, so `$ ` must be compared as `$`.
    let needle = tail.trim_end();
    if needle.is_empty() {
        return prompt_signal(
            "abstain",
            "prompt tail is whitespace only",
            Some(tail),
            Some(&last),
        );
    }
    let occurrences = last.match_indices(needle).count();
    if occurrences == 1 && last.trim_end().ends_with(needle) {
        return prompt_signal(
            "clean",
            &format!("last line ends with prompt tail {needle:?}"),
            Some(tail),
            Some(&last),
        );
    }
    if occurrences > 0 {
        let excerpt = last.trim_end().chars().rev().take(80).collect::<String>();
        let excerpt = excerpt.chars().rev().collect::<String>();
        let reason = if occurrences > 1 {
            format!(
                "prompt tail {needle:?} occurs {occurrences} times; possible text follows an earlier prompt: {excerpt:?}"
            )
        } else {
            format!("text typed after prompt tail {needle:?}: {excerpt:?}")
        };
        return prompt_signal("dirty", &reason, Some(tail), Some(&last));
    }
    prompt_signal(
        "abstain",
        &format!("prompt tail {needle:?} not found in last line"),
        Some(tail),
        Some(&last),
    )
}

/// Take one readiness reading without sleeping or retrying.
pub fn assess<A: HerdrApi + ?Sized>(
    client: &A,
    pane_id: &str,
    config: &Config,
    prompt_tail: Option<&str>,
    read_lines: usize,
) -> Result<Readiness> {
    let process = assess_process(&client.process_info(pane_id)?, config);
    let prompt = if config.readiness == "process" {
        prompt_signal(
            "abstain",
            "prompt check disabled (readiness: process)",
            prompt_tail,
            None,
        )
    } else {
        assess_prompt(
            &client.read(pane_id, "recent-unwrapped", Some(read_lines))?,
            prompt_tail,
        )
    };
    let (ready, reason) = if !process.idle {
        (false, "pane is busy")
    } else if prompt.verdict == "dirty" {
        (false, "pane has an unsubmitted command line")
    } else {
        (true, "pane is idle at a shell prompt")
    };
    Ok(Readiness {
        ready,
        reason: reason.to_owned(),
        process,
        prompt,
    })
}

fn prompt_signal(
    verdict: &str,
    reason: &str,
    tail: Option<&str>,
    last_line: Option<&str>,
) -> PromptSignal {
    PromptSignal {
        verdict: verdict.to_owned(),
        reason: reason.to_owned(),
        tail: tail.map(str::to_owned),
        last_line: last_line.map(str::to_owned),
    }
}

fn zero_width() -> &'static Regex {
    static REGEX: OnceLock<Regex> = OnceLock::new();
    REGEX.get_or_init(|| Regex::new(r"\\\[.*?\\\]").expect("valid zero-width regex"))
}

fn dynamic() -> &'static Regex {
    static REGEX: OnceLock<Regex> = OnceLock::new();
    REGEX.get_or_init(|| {
        Regex::new(r"\\.|\$\([^)]*\)|\$\{[^}]*\}|\$[A-Za-z_][A-Za-z0-9_]*")
            .expect("valid prompt-dynamic regex")
    })
}

fn ps1_assignment() -> &'static Regex {
    static REGEX: OnceLock<Regex> = OnceLock::new();
    REGEX.get_or_init(|| {
        Regex::new(
            r#"(?m)^[ \t]*(?:export[ \t]+)?(?:PS1|PROMPT)=(\"(?:[^\"\\]|\\.)*\"|'[^']*'|[^\s#]+)[ \t]*(?:#.*)?$"#,
        )
        .expect("valid PS1-assignment regex")
    })
}

#[cfg(test)]
mod tests {
    use std::sync::atomic::{AtomicU64, Ordering};

    use crate::client::Pane;
    use crate::error::HerdrRunError;

    use super::*;

    static TEMPORARY_SEQUENCE: AtomicU64 = AtomicU64::new(0);

    fn config() -> Config {
        Config::default()
    }

    fn idle_info(shell: &str) -> ProcessInfo {
        ProcessInfo {
            pane_id: "p1".to_owned(),
            shell_pid: 42,
            foreground_pgid: 42,
            foreground: vec![(42, shell.to_owned(), format!("-{shell}"))],
        }
    }

    #[test]
    fn ps1_parser_ignores_comments_and_last_assignment_wins() {
        let text = r#"
# PS1='ignored> '
PS1='first> '
export PS1="\u@\h:\w\$ " # selected
"#;
        assert_eq!(parse_ps1(text).as_deref(), Some(r"\u@\h:\w\$ "));
        assert_eq!(
            prompt_tail_of(r"\[\033[34m\][\u@\h \w]\n\$ ").as_deref(),
            Some("$ ")
        );
    }

    #[test]
    fn prompt_inference_uses_explicit_value_then_account_files() {
        let mut explicit = config();
        explicit.prompt_tail = Some("PROMPT> ".to_owned());
        assert_eq!(
            infer_prompt_tail(&explicit, Some(Path::new("/does/not/exist"))).as_deref(),
            Some("PROMPT> ")
        );

        let root = std::env::temp_dir().join(format!(
            "herdr-run-readiness-{}-{}",
            std::process::id(),
            TEMPORARY_SEQUENCE.fetch_add(1, Ordering::Relaxed)
        ));
        fs::create_dir_all(&root).unwrap();
        fs::write(root.join(".bashrc"), "PS1='\\u@\\h\\$ '\n").unwrap();
        assert_eq!(
            infer_prompt_tail(&config(), Some(&root)).as_deref(),
            Some("$ ")
        );
        fs::remove_dir_all(root).unwrap();
    }

    #[test]
    fn process_signal_requires_shell_pid_group_and_known_name() {
        assert!(assess_process(&idle_info("bash"), &config()).idle);
        let mut busy_group = idle_info("bash");
        busy_group.foreground_pgid = 99;
        assert!(!assess_process(&busy_group, &config()).idle);
        let mut extra = idle_info("bash");
        extra
            .foreground
            .push((43, "git".to_owned(), "git status".to_owned()));
        assert!(!assess_process(&extra, &config()).idle);
        assert!(!assess_process(&idle_info("python"), &config()).idle);
    }

    #[test]
    fn prompt_tail_is_clean_only_once_at_the_end() {
        assert_eq!(assess_prompt("host\n$\n", Some("$ ")).verdict, "clean");
        assert_eq!(
            assess_prompt("host\n$ git status\n", Some("$ ")).verdict,
            "dirty"
        );
        assert_eq!(
            assess_prompt("host\n$ echo $\n", Some("$ ")).verdict,
            "dirty"
        );
        assert_eq!(
            assess_prompt("host\nunknown\n", Some("$ ")).verdict,
            "abstain"
        );
        assert_eq!(assess_prompt("", None).verdict, "abstain");
    }

    struct ReadinessFake {
        info: ProcessInfo,
        text: String,
    }

    impl HerdrApi for ReadinessFake {
        fn ensure_server(&self) -> Result<bool> {
            unreachable!()
        }
        fn server_running(&self) -> bool {
            unreachable!()
        }
        fn workspace_id_for_label(&self, _label: &str) -> Result<Option<String>> {
            unreachable!()
        }
        fn workspace_label_for_id(&self, _workspace_id: &str) -> Result<Option<String>> {
            unreachable!()
        }
        fn create_workspace(&self, _label: &str, _cwd: &str) -> Result<(String, String, String)> {
            unreachable!()
        }
        fn tab_id_for_label(&self, _workspace_id: &str, _label: &str) -> Result<Option<String>> {
            unreachable!()
        }
        fn create_tab(&self, _workspace_id: &str, _label: &str, _cwd: &str) -> Result<String> {
            unreachable!()
        }
        fn rename_tab(&self, _tab_id: &str, _label: &str) -> Result<()> {
            unreachable!()
        }
        fn panes(&self, _workspace_id: Option<&str>) -> Result<Vec<Pane>> {
            unreachable!()
        }
        fn pane_exists(&self, _pane_id: &str) -> bool {
            unreachable!()
        }
        fn process_info(&self, _pane_id: &str) -> Result<ProcessInfo> {
            Ok(self.info.clone())
        }
        fn read(&self, _pane_id: &str, source: &str, lines: Option<usize>) -> Result<String> {
            assert_eq!(source, "recent-unwrapped");
            assert_eq!(lines, Some(4));
            Ok(self.text.clone())
        }
        fn run(&self, _pane_id: &str, _command: &str) -> Result<()> {
            Err(HerdrRunError::unavailable("unused"))
        }
        fn send_keys(&self, _pane_id: &str, _keys: &str) -> Result<()> {
            unreachable!()
        }
    }

    #[test]
    fn combined_readiness_treats_prompt_as_veto_not_authority() {
        let idle = ReadinessFake {
            info: idle_info("bash"),
            text: "host\nno recognizable prompt\n".to_owned(),
        };
        let reading = assess(&idle, "p1", &config(), Some("$ "), 4).unwrap();
        assert!(reading.ready);
        assert_eq!(reading.prompt.verdict, "abstain");

        let dirty = ReadinessFake {
            info: idle_info("bash"),
            text: "host\n$ half typed\n".to_owned(),
        };
        assert!(
            !assess(&dirty, "p1", &config(), Some("$ "), 4)
                .unwrap()
                .ready
        );

        let mut busy_info = idle_info("bash");
        busy_info.foreground_pgid = 100;
        let busy = ReadinessFake {
            info: busy_info,
            text: "$\n".to_owned(),
        };
        assert!(!assess(&busy, "p1", &config(), Some("$ "), 4).unwrap().ready);
    }

    #[test]
    fn process_only_mode_does_not_read_prompt() {
        let fake = ReadinessFake {
            info: idle_info("bash"),
            text: "$ typed".to_owned(),
        };
        let mut config = config();
        config.readiness = "process".to_owned();
        let reading = assess(&fake, "p1", &config, Some("$ "), 4).unwrap();
        assert!(reading.ready);
        assert_eq!(reading.prompt.verdict, "abstain");
    }
}
